backend fallback test

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30370 [wip] Use codegen'ed unboxing wrappers
* #29934 Boxed variable dispatch
* **#29682 backend fallback test**
* #29681 better boxing
* #29337 Hide the OperatorKernel* argument from the stack based kernel API
* #29330 OperatorHandle::callBoxed/callUnboxed
* #29214 Test cases for backend fallback kernels
* #29201 Add a OperatorHandle argument to boxed kernels
* #30340 Make Dispatcher::backendFallbackKernels_ an array
* #30333 [pytorch][PR] Remove LeftRight from OperatorEntry and DispatchTable.
* #30332 [pytorch][PR] Convert KernelTable to a flat-indexed array rather than a hashtable.

This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

smessmer committed:
backend fallback test

This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/test/backend_fallback_test.cpp:
```diff
@@ -0,0 +1,141 @@
+// TODO Remove callBoxedWorkaround once op.callBoxed(stack) works for all ops.
+//      Once callBoxedWorkaround is removed, we can move this file to the location
+//      where it actually belongs, i.e. next to Dispatcher.h. The only reason for
+//      this not being there yet is that callBoxedWorkaround depends on JIT.
```
You're gonna fix this right? :)

smessmer replied:
yes but it's a long process.

ezyang reviewed (approved):
cc @mcarilli

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

mcarilli commented:
Thanks, this functionality will certainly enable autocasting, at least unless/until a fallthrough mechanism is implemented (proposed by Ed in https://github.com/pytorch/pytorch/issues/28386, updated proposal in https://github.com/pytorch/pytorch/issues/29548).

There are some issues I'm confused about, though:

Is it essential that my fallback function be a "boxed function" and that I pass it to registerBackendFallbackKernel via [makeFromBoxedFunction](https://github.com/pytorch/pytorch/blob/284558f76f5beca401d23b23d06d62ae9728e44c/aten/src/ATen/test/backend_fallback_test.cpp#L119)?  This makes some sense to me because the catchall kernel must be able to handle any arguments, right?

In the current README for op_registration, [your example](https://github.com/pytorch/pytorch/tree/284558f76f5beca401d23b23d06d62ae9728e44c/aten/src/ATen/core/op_registration#catch-all-kernels) shows registering an catchall kernel with arguments manually specified (ie, not wrapped in a generic Stack type or something).  Is that example misleading?  Will it fail as a fallback for ops that accept different arguments?

All I need is fallback function that declares an RAII exclude guard and forwards the arguments to the current op.  For all my non-fallback ops, I also have to cast some of the arguments, but this is straightforward because I register for each of these explicitly, and can have each of my wrapper functions accept and handle the same explicit signature (I wrote a fun metaprogram to handle arbitrary functions as part of my [tentative autocasting PR](https://github.com/pytorch/pytorch/blob/1c4c08d0582f023ae736f4a2964aaa193f64701e/aten/src/ATen/core/autocast/register_autocast_ops.cpp)).  However, for fallback, I need a single catchall function that can serve as a fallback for any op.  Is leaning on the jit operator registry, as your test does, the only way to do so?  Is that what you guys are working on/what Ed was alluding to [here](https://github.com/pytorch/pytorch/pull/29682/files#r345814415)?

smessmer commented:
> Is it essential that my fallback function be a "boxed function" and that I pass it to registerBackendFallbackKernel via [makeFromBoxedFunction](https://github.com/pytorch/pytorch/blob/284558f76f5beca401d23b23d06d62ae9728e44c/aten/src/ATen/test/backend_fallback_test.cpp#L119)? This makes some sense to me because the catchall kernel must be able to handle any arguments, right?

That's correct. You cannot write an unboxed kernel in a way that works with different operator signatures.

> In the current README for op_registration, [your example](https://github.com/pytorch/pytorch/tree/284558f76f5beca401d23b23d06d62ae9728e44c/aten/src/ATen/core/op_registration#catch-all-kernels) shows registering an catchall kernel with arguments manually specified (ie, not wrapped in some generic container). Is that example misleading? Will it fail as a fallback for ops that accept different arguments?

Catch-all kernels and backend-fallback kernels are different things. A catchall kernel is bound to a specific op and called if it's called with a tensor for which this op doesn't have a concrete backend registered. A backend-fallback kernel is not bound to an op but to a backend type and is called when any op is called with a tensor of that backend type.

> All I need is fallback function that declares an RAII exclude guard and forwards the arguments to the current op. It would be nice if a single catchall function could serve as a fallback for any op. Is leaning on the jit operator registry, as your test does, the only way to do so? Is that what you guys are working on/what Ed was alluding to [here](https://github.com/pytorch/pytorch/pull/29682/files#r345814415)?

As mentioned in the comment there, the right way would be `op.callBoxed()` and you should use that whenever it works (i.e. doesn't crash). We'll enable this for more and more ops over time while we fix the system. For the ops that don't work yet, you'll have to lean on the jit operator registry.

> [This backend_fallback_kernel](https://github.com/pytorch/pytorch/blob/8d794700acbd6e8737e92c417f533135a905b2a2/aten/src/ATen/core/op_registration/op_registration_test.cpp#L632-L634) looks promising, but it doesn't actually call into `op`.

That is the same API that this PR uses too.

ezyang commented:
> A catchall kernel is bound to a specific op and called if it's called with a tensor for which this op doesn't have a concrete backend registered

But in #29548 we propose to get rid of it, FYI.

mcarilli commented:
> Catch-all kernels and backend-fallback kernels are different things. A catchall kernel is bound to a specific op and called if it's called with a tensor for which this op doesn't have a concrete backend registered. A backend-fallback kernel is not bound to an op but to a backend type and is called when any op is called with a tensor of that backend type.

That makes sense.  For a second I thought a given catchall kernel was intended _for all ops and TensorTypeIds_.  Thanks for clarifying, I should have looked at the code more closely.

> As mentioned in the comment there, the right way would be op.callBoxed() and you should use that whenever it works (i.e. doesn't crash). We'll enable this for more and more ops over time while we fix the system. For the ops that don't work yet, you'll have to lean on the jit operator registry.

I'd like to write a fallback that correctly accommodates your ongoing work.  Let's say I imitate the op_registration_test.cpp version, [`void backend_fallback_kernel(c10::OperatorKernel*, const c10::OperatorHandle& op, c10::Stack* stack)`](https://github.com/pytorch/pytorch/blob/8d794700acbd6e8737e92c417f533135a905b2a2).  So my function would look like

```c++
void autocast_mode_fallback(c10::OperatorKernel*, const c10::OperatorHandle& op, c10::Stack* stack)
{
  c10::impl::ExcludeTensorTypeIdGuard guard(TensorTypeId::AutocastTensorId);
  op.callBoxed(stack);
}
```
?  `callBoxed` does not appear to be a member of [OperatorHandle](https://github.com/pytorch/pytorch/blob/8d794700acbd6e8737e92c417f533135a905b2a2/aten/src/ATen/core/dispatch/Dispatcher.h#L144-L165) in this PR but I guess you're adding it in [a separate PR](https://github.com/pytorch/pytorch/pull/29330/files).

smessmer commented:
This is the same API but it changed in some of the PRs stacked below this PR.

```
// old API
void autocast_mode_fallback(c10::OperatorKernel*, const c10::OperatorHandle& op, c10::Stack* stack)

// new API
void autocast_mode_fallback(const c10::OperatorHandle& op, c10::Stack* stack)
```

see https://github.com/pytorch/pytorch/pull/29337 and https://github.com/pytorch/pytorch/pull/29201

mcarilli commented:
Thanks!  This will certainly unblock me until fallthrough is available.

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

mcarilli commented on aten/src/ATen/test/backend_fallback_test.cpp:
```diff
@@ -0,0 +1,141 @@
+}
+
+struct Environment {
+  c10::RegistrationHandleRAII registry1 = c10::Dispatcher::singleton().registerBackendFallbackKernel(
```
Afaict all your tests for backend fallback kernels register them by a direct call into `c10::Dispatcher::singleton().registerBackendFallbackKernel()`, as opposed to through `torch::RegisterOperators` that's used for other types of kernels (including catchAllKernels).  Do you plan to expose the backend fallback registration through `torch::RegisterOperators` as well, or will registering directly onto the dispatcher instance continue to be the plan?  It's usable for me either way.

smessmer replied:
this is still under discussion. For now, please use the dispatcher instance directly.

smessmer commented on aten/src/ATen/test/backend_fallback_test.cpp:
```diff
@@ -0,0 +1,141 @@
+}
+
+struct Environment {
+  c10::RegistrationHandleRAII registry1 = c10::Dispatcher::singleton().registerBackendFallbackKernel(
```
yes. Keep it around somewhere, e.g. in a static global.

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]

smessmer committed:
Update on "backend fallback test"


This PR re-introduces backend_fallback_test.cpp, which was previously called boxed_fallback_test.cpp and showed how to use the backend fallback API.

Differential Revision: [D18462654](https://our.internmc.facebook.com/intern/diff/D18462654/)

[ghstack-poisoned]
